ecl/hql/hqlcache.cpp

```diff
@@ -340,6 +342,161 @@ void convertSelectsToPath(StringBuffer & filename, const char * eclPath)
 }
 
 //---------------------------------------------------------------------------------------------------------------------
+static HqlTransformerInfo HqlSimplifiedRecordTransformerInfo("HqlSimplifiedRecordTransformer");
+class HqlSimplifiedRecordTransformer : public QuickHqlTransformer
+{
+public:
+    HqlSimplifiedRecordTransformer() : QuickHqlTransformer(HqlSimplifiedRecordTransformerInfo, nullptr)
+    {
+    }
+
+    virtual IHqlExpression * createTransformed(IHqlExpression * expr)
+    {
+        IHqlExpression * body = expr->queryBody(false);
+        if (expr != body)
+            return transform(body);
+
+        node_operator op = expr->getOperator();
+        switch (op)
+        {
+        case no_record:
+            {
+                HqlExprArray children;
+                bool same = true;
+                ForEachChild(idx, expr)
+                {
+                    IHqlExpression * cur = expr->queryChild(idx);
+                    if (!cur->isAttribute())
+                    {
+                        OwnedHqlExpr mapped = transform(cur);
+                        if (!mapped)
+                            throw MakeStringExceptionDirect(0, "Cannot simplify record");
+                        if (mapped != cur)
+                            same = false;
+                        children.append(*(mapped.getClear()));
+
+                    }
+                    else
+                        same = false;
+                }
+                if (same)
+                    return LINK(expr);
+                return expr->clone(children);
+            }
+        case no_field:
+            {
+                ITypeInfo * type = expr->queryType();
+                if (type->getTypeCode()==type_alien)
+                    throw MakeStringExceptionDirect(0, "Cannot simplify record");
+                HqlExprArray children;
+                bool same = true;
+                ForEachChild(idx, expr)
+                {
+                    IHqlExpression * cur = expr->queryChild(idx);
+                    if (cur->isAttribute())
+                    {
+                        if (cur->queryName()==_implicitFpos_Atom)
```

Comment:
Should have a comment explaining why this is special-cased.

Reply:
Still hesitant about this.  I would merge it with the code included, but there should be a jira to revisit it - and investigate how many attributes it affects.
Currently if the records do not match then a simplified expression is not created - which means that almost all of the other code is not actually needed - but would be if that condition was relaxed.